### DISCUSSION TITLE
[Discussion] Media Management for FPP

### DISCUSSION THREAD

**Discussion by kylegrymonprez** [xLightsSequencer/xLights#4855] [Ideas]
**Is your feature request related to a problem? Please describe.**
One of the common things people are doing is uploading a video file to remote FPP instances in order to 'trick' HDMI out using the full video vs sequence data on that remote pi.

**Describe the solution you'd like**
Should we add some sort of multiple media management abilities to xLights and FPP Connect in order to manage which media files should do to which FPP devices? I think it could be interesting.

**Comment by cybercop23** [comment]
That's already there. Create a folder with the same exact FPP name in your media folder and put the MP4 with the same exact name as the sequence mp3 and check media upload in FPP connect. It will first go through this process to find act matching items and load them to the appropriate FPP.
Attentively you can also rename the media/MP4 and add the same exact FPP name to the end of the filename and will have the same effect as above. But I think having a separate folder per FPP instance if better file management.

**Reply by kylegrymonprez** [reply]
I'm thinking something more like "Matrix-1_video" goes to "FPPMatrix-1.local" and "Matrix-2_video" goes to "FPPMatrixs-2.local", but during the media upload it renames the .mp4 accordingly for each host.

**Comment by derwin12** [comment]
Can we make this easier? Perhaps a placeholder effect in the sequence? Just tossing out ideas..

**Reply by kylegrymonprez** [reply]
that would need FPP to parse the sequence data to understand the mappings.

**Comment by cybercop23** [comment]
They have to manage and create the content, assuming it's not just the sequence media, if it is, checking that box will just upload it. You don't want FPP connect to search through the sequence file to find it and then there's multiple VMs or even split content among left/right projector.

**Comment by kylegrymonprez** [comment]
yeah - i'm thinking more like file X goes to FPP instance A, file B goes to FPP instance B and FPP connect takes all of the file naming and stuff into account on it's own. Just a mapping of file on compter to FPP instance, and the upload process handles the file naming stuff.

**Comment by cybercop23** [comment]
and why not use Discussions... :)

**Comment by kylegrymonprez** [comment]
(oops forgot about discussions - I think someone can move it over - I can't)

**Comment by cybercop23** [comment]
I'm lost.. that's what it does now as option B above.

**Reply by kylegrymonprez** [reply]
Are you talking about the folder structure thing? Is this documented somewhere?

**Reply by cybercop23** [reply]
> Are you talking about the folder structure thing? Is this documented somewhere?

In my desc.. it lists option B.. file name.. MyBigMatrixVideoFPPVirtualMatrix.mp4.. it will send it to FPPVirtualMatrix and strip fpp's description from the filename.

**Comment by derwin12** [comment]
Can we put this in xlights some how/some where? Nobody knows anything about file management for windows :-)

**Reply by kylegrymonprez** [reply]
We may be able to better highlight this as well in the FPP Connect. Poking around the code I think I see where this is happening [here](https://github.com/xLightsSequencer/xLights/blob/fd4f2807e5b718cfdc21d6794ee145cf5ee3cf6f/xLights/controllers/FPP.cpp#L955).

What if we attempt to traverse this check into the FPP Connect:
During the sequence population, see if there are any files of "Host Specific Media" that exist for the listed FPP hosts in the dialog above. This would allow us to at least signifiy in the dialog that FPP Connect will be pushing multiple media files to different hosts if the Media checkbox is enabled AND the FPP instance is checked. 
- This would also mean we dont have to make the dialog reactive to checkboxes getting marked.

**Comment by kylegrymonprez** [comment]
After further discussion we're proposing a two-fold change:

1 - In "Sequence Settings, add another tab for "Alternate Media" or "Other Media". On this pane, you would map FPP Hostname to a different media file than what's on the 'main' sequence media. 

- This would become part of the FSEQ header file so that FPP Connect could read and understand it without scanning the file past the header. This needs to be backwards compatible for load, but a forward looking feature.
![Screenshot 2024-10-02 at 2 32 42 PM](https://github.com/user-attachments/assets/370725ae-df9a-45d3-9856-f9c8e2311d03)


2 - The "FPP Connect" Dialog would need to be updated to display the media, and also allow for some further options for media. We're proposing instead of a checkbox, a dropdown similar to "Models" where the options are "None", "Seq"uence, and "Other"/"Alternate".
![Screenshot 2024-10-02 at 2 35 12 PM](https://github.com/user-attachments/assets/9e0f34ba-34db-4ab1-add4-b649594a0dd5)

**Comment by kylegrymonprez** [comment]
OOO /api/system/info has an entry for the FPP UUID already there - @computergeek1507 would you have any issue if instead of IP or Hostname, we store settings to the computer by UUID? This would also address the issues people encounter when using hostname intead of static IPs for their FPP Connect selections.

**Reply by kylegrymonprez** [reply]
Proposed hierarchy: UUID > IP > hostname. There are tradeoffs either way depending on the ordering. At least this one remains consistent with the current hierarchy.

**Reply by cybercop23** [reply]
I'd say... UUID, Hostname and IP

**Reply by kylegrymonprez** [reply]
Actually - the 'hidden' media mapping system is by hostname - so the whole IP vs Hostname thing is a bit up in the air with regards to 'which one should take precedent' - so I could go either way.

**Comment by AGFazio** [comment]
Could this be done with adding a check box/drop down in the video effect settings window? 

If someone was to click the check box/drop down, xlights would then skip the video file from the render (to save render time) and prepare it for the FPP instance it would go to as the model is set to an FPP instance in the visualizer.  This prep would either add the FPP instance name to the end of the file for media upload or put the file in the media/FPP instance folder automagically.
